src/webgpu/util/shader.ts

```diff
@@ -0,0 +1,112 @@
+import { unreachable } from '../../common/util/util.js';
+
+/**
+ * Build a fragment shader based on output value and types
+ * e.g. write to color target 0 a vec4<f32>(1.0, 0.0, 1.0, 1.0) and color target 2 a vec2<u32>(1, 2)
+ * outputs: [
+ *   {
+ *     values: [1, 0, 1, 1],,
+ *     sampleType: 'float',
+ *     componentCount: 4,
+ *   },
+ *   null,
+ *   {
+ *     values: [1, 2],
+ *     sampleType: 'uint',
+ *     componentCount: 2,
+ *   },
+ * ]
+ *
+ * return:
+ * struct Outputs {
+ *     @location(0) o1 : vec4<f32>;
+ *     @location(2) o3 : vec2<u32>;
+ * }
+ * @stage(fragment) fn main() -> Outputs {
+ *     return Outputs(vec4<f32>(1.0, 0.0, 1.0, 1.0), vec4<u32>(1, 2));
+ * }
+ * @param outputs the shader outputs for each location attribute
+ * @returns the fragment shader string
+ */
+export function getFragmentShaderCodeWithOutput(
+  outputs: ({
+    values: readonly number[];
+    sampleType: GPUTextureSampleType;
```

Comment:
odd that an input is GPUTextureSampleType when no sampling is giong on.

Perhaps it should take in `i32 | u32 | f32` and then have a separate helper that maps GPUTextureSampleType to that union